packages/headless/src/controllers/url-hash-manager/headless-url-hash-manager.ts

```diff
@@ -0,0 +1,196 @@
+import {Schema, StringValue} from '@coveo/bueno';
+import {Engine} from '../../app/headless-engine';
+import {
+  restoreSearchParameters,
+  SearchParameters,
+} from '../../features/search-parameters/search-parameter-actions';
+import {SearchParametersState} from '../../state/search-app-state';
+import {validateInitialState} from '../../utils/validate-payload';
+import {buildController, Controller} from '../controller/headless-controller';
+import {executeSearch} from '../../features/search/search-actions';
+import {ConfigurationSection} from '../../state/state-sections';
+import {logSearchboxSubmit} from '../../features/query/query-analytics-actions';
+import {
+  AnalyticsType,
+  makeNoopAnalyticsAction,
+} from '../../features/analytics/analytics-utils';
+import {logResultsSort} from '../../features/sort-criteria/sort-criteria-analytics-actions';
+import {
+  logFacetClearAll,
+  logFacetDeselect,
+  logFacetSelect,
+} from '../../features/facets/facet-set/facet-set-analytics-actions';
+import {buildSearchParameterSerializer} from '../../features/search-parameters/search-parameter-serializer';
+import {
+  buildSearchParameterManager,
+  getInitialSearchParameterState,
+} from '../search-parameter-manager/headless-search-parameter-manager';
+
+export interface UrlHashManagerProps {
+  /**
+   * The initial state that should be applied to the `UrlHashManager` controller.
+   */
+  initialState: UrlHashManagerInitialState;
+}
+
+export interface UrlHashManagerInitialState {
+  /**
+   * The url hash containing the parameters affecting the search response.
+   */
+  urlHash: string;
+}
+
+const urlHashValue = new StringValue({regex: /^#/});
+const initialStateSchema = new Schema<Required<UrlHashManagerInitialState>>({
+  urlHash: urlHashValue,
+});
+
+/**
+ * The `UrlHashManager` controller allows managing an url hash which affect the search response.
+ * */
+export interface UrlHashManager extends Controller {
+  /**
+   * The state relevant to the `UrlHashManager` controller.
+   * */
+  state: UrlHashManagerState;
+  /**
+   * Updates the search parameters from the url hash & launches a search.
+   * @param urlHash The url hash containing the parameters affecting the search.
+   */
+  updateUrlHash(urlHash: string): void;
+}
+
+export interface UrlHashManagerState {
+  /**
+   * The url hash containing the parameters affecting the search response.
+   */
+  urlHash: string;
+}
+
+/**
+ * Creates a `UrlHashManager` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `UrlHashManager` properties.
+ * @returns A `UrlHashManager` controller instance.
+ */
+export function buildUrlHashManager(
```

Comment:
Could it be just named URL manager ? 

Does it 100% *need* to be used with the hash portion of the URL ?

If not, maybe the name is too restrictive ?

Said differently, if somebody wanted to use query string parameter to store their state, would we really need to create a `buildUrlQueryStringManager` ?